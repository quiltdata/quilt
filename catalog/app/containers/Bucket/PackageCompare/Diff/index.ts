export { default as Entries } from './Entries'
export { default as Metadata } from './Metadata'
export { default as Summary } from './Summary'
