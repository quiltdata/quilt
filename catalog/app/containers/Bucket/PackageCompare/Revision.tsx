import * as dateFns from 'date-fns'
import * as React from 'react'
import * as M from '@material-ui/core'

const useDetailsStyles = M.makeStyles((t) => ({
  message: {
    display: 'block',
  },
  hash: {
    ...t.typography.monospace,
    display: 'block',
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
  },
}))

interface DetailsProps {
  message: string | null
  hash: string
}

export function Details({ message, hash }: DetailsProps) {
  const classes = useDetailsStyles()
  return (
    <>
      <span className={classes.message}>{message || 'No message'}</span>
      <span className={classes.hash}>{hash}</span>
    </>
  )
}

interface DateProps {
  modified: Date
}

export function Date({ modified }: DateProps) {
  return <>{dateFns.format(modified, 'MMM d yyyy - h:mma')}</>
}
