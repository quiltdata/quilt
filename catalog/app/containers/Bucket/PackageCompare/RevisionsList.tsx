import * as dateFns from 'date-fns'
import * as React from 'react'
import type { ResultOf } from '@graphql-typed-document-node/core'
import * as M from '@material-ui/core'

import * as GQL from 'utils/GraphQL'

import REVISION_LIST_QUERY from './gql/RevisionList.generated'

type RevisionFields = NonNullable<
  NonNullable<
    ResultOf<typeof REVISION_LIST_QUERY>['package']
  >['revisions']['page'][number]
>

const useStyles = M.makeStyles((t) => ({
  root: {
    overflow: 'hidden',
  },
  hash: {
    fontFamily: t.typography.monospace.fontFamily,
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
  },
}))

interface RevisionsListProps {
  packageHandle: { bucket: string; name: string }
  value: string
  onChange: (hash: string) => void
  temporaryRemoveNone?: boolean
}

export default function RevisionsList({
  temporaryRemoveNone = false,
  packageHandle,
  onChange,
  value,
}: RevisionsListProps) {
  const classes = useStyles()
  const revisionListQuery = GQL.useQuery(REVISION_LIST_QUERY, {
    bucket: packageHandle.bucket || '',
    name: packageHandle.name || '',
    page: 1,
    perPage: 100, // Get enough revisions for the dropdown
  })

  const handleChange = (event: React.ChangeEvent<{ value: unknown }>) => {
    onChange(event.target.value as string)
  }

  return (
    <M.Select
      value={value}
      onChange={handleChange}
      displayEmpty
      disabled={revisionListQuery.fetching}
      fullWidth
      className={classes.root}
    >
      {!temporaryRemoveNone && (
        <M.MenuItem value="">
          <em>None</em>
        </M.MenuItem>
      )}
      {GQL.fold(revisionListQuery, {
        data: (d) =>
          d.package?.revisions.page.map((r: RevisionFields) => (
            <M.MenuItem key={r.hash} value={r.hash}>
              <M.ListItemText
                primary={dateFns.format(r.modified, 'MMM d yyyy - h:mma')}
                secondary={
                  <>
                    {r.message || 'No message'}
                    <br />
                    <span className={classes.hash}>{r.hash}</span>
                  </>
                }
              />
            </M.MenuItem>
          )) || [],
        error: () => [
          <M.MenuItem key="error" disabled>
            Error loading revisions
          </M.MenuItem>,
        ],
        fetching: () => [
          <M.MenuItem key="loading" disabled>
            Loading revisions...
          </M.MenuItem>,
        ],
      })}
    </M.Select>
  )
}
