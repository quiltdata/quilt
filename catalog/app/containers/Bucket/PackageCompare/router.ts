import invariant from 'invariant'
import * as React from 'react'
import * as RRDom from 'react-router-dom'

import * as NamedRoutes from 'utils/NamedRoutes'
import type { PackageHandle } from 'utils/packageHandle'
import parseSearch from 'utils/parseSearch'

interface PackageCompareParams {
  bucket: string
  name: string
  baseHash: string
  otherHash: string
}

type Single = [PackageHandle]
type Pair = [PackageHandle, PackageHandle]

export const isPair = (x: Single | Pair): x is Pair => x.length > 1

export function useRouter() {
  const { bucket, name, baseHash, otherHash } = RRDom.useParams<PackageCompareParams>()

  invariant(!!bucket, '`bucket` must be defined')
  invariant(!!name, '`name` must be defined')
  invariant(!!baseHash, '`baseHash` must be defined')

  const { push } = RRDom.useHistory()
  const { urls } = NamedRoutes.use()
  const location = RRDom.useLocation()
  const { showAll } = parseSearch(location.search)

  const pair: Single | Pair = React.useMemo(() => {
    const base = { bucket, name, hash: baseHash }
    if (!otherHash) return [base]
    return [base, { bucket, name, hash: otherHash }]
  }, [bucket, name, baseHash, otherHash])

  const changeBase = React.useCallback(
    (hash: string) =>
      push(urls.bucketPackageCompare(bucket, name, hash, otherHash, { showAll })),
    [bucket, name, push, otherHash, urls, showAll],
  )
  const changeOther = React.useCallback(
    (hash: string) =>
      push(urls.bucketPackageCompare(bucket, name, baseHash, hash, { showAll })),
    [bucket, name, push, baseHash, urls, showAll],
  )
  const swap = React.useCallback(() => {
    invariant(!!otherHash, '`otherHash` must be defined')
    push(urls.bucketPackageCompare(bucket, name, otherHash, baseHash, { showAll }))
  }, [bucket, name, push, baseHash, otherHash, urls, showAll])

  const changesOnly = !showAll || showAll === 'false'
  const toggleChangesOnly = React.useCallback(
    (checked: boolean) =>
      push(
        urls.bucketPackageCompare(bucket, name, baseHash, otherHash, {
          showAll: !checked || null,
        }),
      ),
    [bucket, name, baseHash, otherHash, push, urls],
  )

  return {
    bucket,
    name,

    pair,

    changeBase,
    changeOther,
    swap,

    changesOnly,
    toggleChangesOnly,
  }
}
