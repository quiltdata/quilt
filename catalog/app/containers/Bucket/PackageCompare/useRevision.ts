import * as React from 'react'

import * as Model from 'model'
import * as GQL from 'utils/GraphQL'

import REVISION_QUERY from './gql/Revision.generated'

export type Revision = Pick<
  Model.GQLTypes.PackageRevision,
  'hash' | 'modified' | 'message' | 'userMeta' | 'totalBytes' | 'contentsFlatMap'
>

export type RevisionResult =
  | { _tag: 'loading' }
  | { _tag: 'error'; error: Error }
  | { _tag: 'ok'; revision: Revision }

export function useRevision(
  bucket: string,
  name: string,
  hashOrTag: string,
): RevisionResult {
  const query = GQL.useQuery(REVISION_QUERY, { bucket, name, hashOrTag })

  return React.useMemo(
    () =>
      GQL.fold(query, {
        fetching: () => ({ _tag: 'loading' }),
        error: (error) => ({ _tag: 'error', error }),
        data: (data) =>
          data.package?.revision
            ? { _tag: 'ok', revision: data.package?.revision }
            : { _tag: 'error', error: new Error('No revision data found') },
      }),
    [query],
  )
}
