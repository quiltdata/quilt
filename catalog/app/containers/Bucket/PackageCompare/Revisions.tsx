import * as React from 'react'
import * as M from '@material-ui/core'
import * as Icons from '@material-ui/icons'
import * as Lab from '@material-ui/lab'

import type { PackageHandle } from 'utils/packageHandle'

import useRevisions from './useRevisions'
import RevisionsList from './RevisionsList'

const useHeaderStyles = M.makeStyles((t) => ({
  root: {
    alignItems: 'center',
    display: 'flex',
  },
  revisions: {
    display: 'flex',
    flexDirection: 'column',
    flexGrow: 1,
    gap: t.spacing(1),
  },
  swap: {
    marginLeft: t.spacing(3),
    flexShrink: 0,
  },
}))

interface HeaderProps {
  left: PackageHandle
  right: PackageHandle | null
  onLeftChange: (hash: string) => void
  onRightChange: (hash: string) => void
  onSwap: () => void
}

export default function Revisions({
  left,
  right,
  onLeftChange,
  onRightChange,
  onSwap,
}: HeaderProps) {
  const classes = useHeaderStyles()
  const data = useRevisions(left.bucket, left.name)
  switch (data._tag) {
    case 'loading':
      return (
        <div className={classes.root}>
          <div className={classes.revisions}>
            <Lab.Skeleton width="100%" variant="rect">
              <M.Select fullWidth />
            </Lab.Skeleton>
            <Lab.Skeleton width="100%" variant="rect">
              <M.Select fullWidth />
            </Lab.Skeleton>
          </div>
          <Lab.Skeleton className={classes.swap} variant="circle">
            <M.IconButton />
          </Lab.Skeleton>
        </div>
      )
    case 'error':
      return <Lab.Alert severity="error">{data.error.message}</Lab.Alert>
    case 'ok':
      return (
        <div className={classes.root}>
          <div className={classes.revisions}>
            <RevisionsList
              revisions={data.revisions}
              value={left.hash}
              onChange={onLeftChange}
              temporaryRemoveNone
            />
            <RevisionsList
              revisions={data.revisions}
              value={right?.hash || ''}
              onChange={onRightChange}
            />
          </div>
          <M.IconButton className={classes.swap} onClick={onSwap} disabled={!right}>
            <Icons.SwapVert />
          </M.IconButton>
        </div>
      )
  }
}
