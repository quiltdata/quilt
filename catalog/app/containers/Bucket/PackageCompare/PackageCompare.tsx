import * as React from 'react'
import * as M from '@material-ui/core'

import MetaTitle from 'utils/MetaTitle'
import * as NamedRoutes from 'utils/NamedRoutes'
import StyledLink from 'utils/StyledLink'
import type { PackageHandle } from 'utils/packageHandle'

import * as FileView from '../FileView'
import WithPackagesSupport from '../WithPackagesSupport'

import * as Diff from './Diff'
import RevisionsRange from './RevisionsRange'
import useRevisions from './useRevisionsPair'
import { useRouter, isPair } from './router'

interface PackageNameProps {
  bucket: string
  name: string
}

function PackageName({ bucket, name }: PackageNameProps) {
  const { urls } = NamedRoutes.use()
  return (
    <M.Typography variant="h6">
      Comparing changes in{' '}
      <StyledLink to={urls.bucketPackageDetail(bucket, name)}>{name}</StyledLink>{' '}
      revisions
    </M.Typography>
  )
}

interface ChangesOnlyCheckboxProps {
  className: string
  value: boolean
  onChange: (checked: boolean) => void
}

function ChangesOnlyCheckbox({ className, onChange, value }: ChangesOnlyCheckboxProps) {
  const handleChange = React.useCallback(
    (_event: React.ChangeEvent<HTMLInputElement>, checked) => onChange(checked),
    [onChange],
  )
  return (
    <M.FormControlLabel
      className={className}
      control={<M.Checkbox checked={value} onChange={handleChange} />}
      label="Show changes only"
    />
  )
}

const useStyles = M.makeStyles((t) => ({
  root: {
    marginTop: t.spacing(2),
  },
  checkbox: {
    marginRight: 0,
  },
  table: {
    marginTop: t.spacing(1),
  },
  summary: {
    marginTop: t.spacing(4),
  },
  details: {
    marginTop: t.spacing(4),
    display: 'flex',
    justifyContent: 'space-between',
  },
  userMeta: {},
  entries: {
    marginTop: t.spacing(4),
  },
}))

interface RevisionsCompareProps {
  bucket: string
  name: string
  pair: [PackageHandle, PackageHandle]
  changesOnly: boolean
  onChangesOnly: (changesOnly: boolean) => void
  onBaseChange: (hash: string) => void
  onOtherChange: (hash: string) => void
  onSwap: () => void
}

export function RevisionsCompare({
  bucket,
  name,
  pair,
  changesOnly,
  onChangesOnly,
  onBaseChange,
  onOtherChange,
  onSwap,
}: RevisionsCompareProps) {
  const classes = useStyles()

  const revisionsResult = useRevisions(pair)

  return (
    <div className={classes.root}>
      <RevisionsRange
        bucket={bucket}
        name={name}
        pair={pair}
        onBaseChange={onBaseChange}
        onOtherChange={onOtherChange}
        onSwap={onSwap}
      />

      <div className={classes.summary}>
        <M.Typography variant="h6" gutterBottom>
          What's changed
        </M.Typography>
        <Diff.Summary revisionsResult={revisionsResult} />
      </div>

      <M.Typography variant="h6" gutterBottom className={classes.details}>
        Details
        <ChangesOnlyCheckbox
          className={classes.checkbox}
          value={changesOnly}
          onChange={onChangesOnly}
        />
      </M.Typography>

      <div className={classes.userMeta}>
        <M.Typography variant="subtitle1" gutterBottom>
          User metadata
        </M.Typography>
        <Diff.Metadata revisionsResult={revisionsResult} changesOnly={changesOnly} />
      </div>

      <div className={classes.entries}>
        <M.Typography variant="subtitle1" gutterBottom>
          Entries
        </M.Typography>
        <Diff.Entries revisionsResult={revisionsResult} changesOnly={changesOnly} />
      </div>
    </div>
  )
}

export default function PackageCompareWrapper() {
  const {
    bucket,
    name,

    pair,

    changeBase,
    changeOther,
    swap,

    changesOnly,
    toggleChangesOnly,
  } = useRouter()

  return (
    <>
      <MetaTitle>{[`Comparing changes in ${name} revisions`, bucket]}</MetaTitle>
      <WithPackagesSupport bucket={bucket}>
        <FileView.Root>
          <PackageName bucket={bucket} name={name} />
          {isPair(pair) ? (
            <RevisionsCompare
              bucket={bucket}
              name={name}
              changesOnly={changesOnly}
              pair={pair}
              onBaseChange={changeBase}
              onOtherChange={changeOther}
              onChangesOnly={toggleChangesOnly}
              onSwap={swap}
            />
          ) : (
            <RevisionsRange
              bucket={bucket}
              name={name}
              pair={pair}
              onBaseChange={changeBase}
              onOtherChange={changeOther}
              onSwap={swap}
            />
          )}
        </FileView.Root>
      </WithPackagesSupport>
    </>
  )
}
