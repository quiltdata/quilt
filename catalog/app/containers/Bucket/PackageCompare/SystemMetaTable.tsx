import cx from 'classnames'
import * as dateFns from 'date-fns'
import * as React from 'react'
import * as M from '@material-ui/core'
import * as Lab from '@material-ui/lab'

import * as NamedRoutes from 'utils/NamedRoutes'
import StyledLink from 'utils/StyledLink'
import assertNever from 'utils/assertNever'
import type { PackageHandle } from 'utils/packageHandle'
import { readableBytes } from 'utils/string'

import { Revision, RevisionResult } from './useRevision'
import * as Diff from './Diff'

interface ModifiedProps {
  packageHandle: PackageHandle
  revision: Revision
}

function Modified({
  packageHandle: { bucket, name },
  revision: { modified, hash },
}: ModifiedProps) {
  const { urls } = NamedRoutes.use()
  const formatted = React.useMemo(
    () => dateFns.format(modified, 'MMMM do yyyy - h:mma'),
    [modified],
  )
  return (
    <M.Typography variant="body2">
      <StyledLink to={urls.bucketPackageTree(bucket, name, hash)}>{formatted}</StyledLink>
    </M.Typography>
  )
}

function ModifiedSkeleton() {
  return (
    <M.Typography variant="body2">
      <Lab.Skeleton width={120} />
    </M.Typography>
  )
}

const useMessageStyles = M.makeStyles((t) => ({
  empty: {
    color: t.palette.text.secondary,
    fontStyle: 'italic',
  },
}))

interface MessageProps {
  revision: Revision
}

function Message({ revision: { message } }: MessageProps) {
  const classes = useMessageStyles()
  return (
    <M.Typography
      className={cx(!message && classes.empty)}
      variant="body2"
      component="span"
    >
      {message || 'No message'}
    </M.Typography>
  )
}

function MessageSkeleton() {
  return (
    <M.Typography variant="body2">
      <Lab.Skeleton width={240} />
    </M.Typography>
  )
}

interface SizeProps {
  revision: Revision
}

function Size({ revision: { totalBytes } }: SizeProps) {
  return (
    <M.Typography variant="body2" component="span">
      <M.Tooltip title={`${totalBytes} B`}>{readableBytes(totalBytes)}</M.Tooltip>
    </M.Typography>
  )
}

function SizeSkeleton() {
  return (
    <M.Typography variant="body2">
      <Lab.Skeleton width={60} />
    </M.Typography>
  )
}

const useColumnStyles = M.makeStyles((t) => ({
  cell: {
    padding: t.spacing(1, 0),
    whiteSpace: 'nowrap',
    '& + &': {
      borderTop: `1px solid ${t.palette.divider}`,
    },
  },
}))

interface ColumnProps {
  packageHandle: PackageHandle
  revision: Revision
  other?: Revision
  order: Diff.Order
}

function Column({ packageHandle, revision, other, order }: ColumnProps) {
  const classes = useColumnStyles()
  return (
    <div>
      <div className={classes.cell}>
        <Modified packageHandle={packageHandle} revision={revision} />
      </div>
      <div className={classes.cell}>
        {!other || revision.message == other.message ? (
          <Message revision={revision} />
        ) : (
          <Diff.Change order={order}>
            <Message revision={revision} />
          </Diff.Change>
        )}
      </div>
      <div className={classes.cell}>
        {!other || revision.totalBytes === other.totalBytes ? (
          <Size revision={revision} />
        ) : (
          <Diff.Change order={order}>
            <Size revision={revision} />
          </Diff.Change>
        )}
      </div>
    </div>
  )
}

function ColumnSkeleton() {
  const classes = useColumnStyles()
  return (
    <div>
      <div className={classes.cell}>
        <ModifiedSkeleton />
      </div>
      <div className={classes.cell}>
        <MessageSkeleton />
      </div>
      <div className={classes.cell}>
        <SizeSkeleton />
      </div>
    </div>
  )
}

interface ColumnErrorProps {
  error: Error
}

function ColumnError({ error }: ColumnErrorProps) {
  return <Lab.Alert severity="error">{error.message}</Lab.Alert>
}

interface ColumnWrapperProps {
  packageHandle: PackageHandle
  result: RevisionResult
  other: RevisionResult
}

function ColumnWrapper({ packageHandle, result, other }: ColumnWrapperProps) {
  const order: Diff.Order = React.useMemo(() => {
    if (result._tag !== 'ok' || other._tag !== 'ok') return 'limbo' as const
    return result.revision.modified > other.revision.modified ? 'latter' : 'former'
  }, [result, other])

  switch (result._tag) {
    case 'idle':
      return null
    case 'loading':
      return <ColumnSkeleton />
    case 'error':
      return <ColumnError error={result.error} />
    case 'ok':
      return other._tag === 'ok' ? (
        <Column
          packageHandle={packageHandle}
          revision={result.revision}
          other={other.revision}
          order={order}
        />
      ) : (
        <Column packageHandle={packageHandle} revision={result.revision} order={order} />
      )
    default:
      assertNever(result)
  }
}

const useStyles = M.makeStyles((t) => ({
  grid: {
    display: 'grid',
    gridTemplateColumns: '1fr 1fr',
    gridColumnGap: t.spacing(8),
    marginTop: t.spacing(1),
  },
}))

interface SystemMetaTableProps {
  left: PackageHandle
  right: PackageHandle | null
  leftRevision: RevisionResult
  rightRevision: RevisionResult
}

export default function SystemMetaTable({
  left,
  right,
  leftRevision,
  rightRevision,
}: SystemMetaTableProps) {
  const classes = useStyles()

  return (
    <div className={classes.grid}>
      <ColumnWrapper other={rightRevision} packageHandle={left} result={leftRevision} />
      {right && (
        <ColumnWrapper
          other={leftRevision}
          packageHandle={right}
          result={rightRevision}
        />
      )}
    </div>
  )
}
