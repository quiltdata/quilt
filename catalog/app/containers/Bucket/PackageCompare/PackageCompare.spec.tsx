import * as React from 'react'
import renderer from 'react-test-renderer'

import { RevisionsCompare } from './PackageCompare'

jest.mock('@material-ui/core', () => ({
  ...jest.requireActual('@material-ui/core'),
  makeStyles: () => () => ({}),
}))

// Mock config
jest.mock('constants/config', () => ({}))

// Mock GraphQL hook
jest.mock('utils/GraphQL', () => ({
  useQuery: jest.fn((_query, variables) => {
    // Return different mock data based on the query type
    if (variables.max !== undefined) {
      // This is a manifest query
      return {
        data: {
          package: {
            revision: {
              contentsFlatMap: {
                'file1.txt': { size: 100 },
                'file2.txt': { size: 200 },
              },
            },
          },
        },
      }
    }
    // This is a revision list query
    return {
      data: {
        package: {
          revisions: {
            page: [
              {
                hash: 'hash1',
                modified: new Date('2023-01-01'),
                message: 'Test message 1',
                userMeta: { test: 'meta1' },
                totalBytes: 1000,
                totalEntries: 5,
              },
              {
                hash: 'hash2',
                modified: new Date('2023-01-02'),
                message: 'Test message 2',
                userMeta: { test: 'meta2' },
                totalBytes: 2000,
                totalEntries: 10,
              },
            ],
          },
        },
      },
    }
  }),
  fold: jest.fn((query, cases) => cases.data(query.data)),
}))

// Mock NamedRoutes
jest.mock('utils/NamedRoutes', () => ({
  use: jest.fn(() => ({
    urls: {
      bucketPackageTree: (bucket: string, name: string, hash: string) =>
        `/b/${bucket}/packages/${name}/tree/${hash}`,
      bucketPackageDetail: (bucket: string, name: string) =>
        `/b/${bucket}/packages/${name}`,
    },
  })),
}))

// Mock copyToClipboard
jest.mock('utils/clipboard', () => jest.fn())

// Mock JsonDisplay component
jest.mock('components/JsonDisplay', () =>
  jest.fn(({ value }: { value: any }) => (
    <div data-testid="json-display">{JSON.stringify(value)}</div>
  )),
)

// Mock Skeleton component
jest.mock('components/Skeleton', () => jest.fn(() => <div data-testid="skeleton" />))

// Mock StyledLink component
jest.mock('utils/StyledLink', () =>
  jest.fn(({ to, children }: { to: string; children: React.ReactNode }) => (
    <a href={to}>{children}</a>
  )),
)

// Mock react-diff-viewer-continued
jest.mock('react-diff-viewer-continued', () =>
  jest.fn(
    ({
      oldValue,
      newValue,
      leftTitle,
      rightTitle,
    }: {
      oldValue: string
      newValue: string
      leftTitle: string
      rightTitle: string
    }) => (
      <div data-testid="diff-viewer">
        <div>{leftTitle}</div>
        <div>{rightTitle}</div>
        <pre>{oldValue}</pre>
        <pre>{newValue}</pre>
      </div>
    ),
  ),
)

describe('containers/Bucket/PackageCompare/PackageCompare', () => {
  const mockPackageHandles = {
    left: {
      bucket: 'test-bucket',
      name: 'test-package',
      hash: 'hash1',
    },
    right: {
      bucket: 'test-bucket',
      name: 'test-package',
      hash: 'hash2',
    },
    onLeftChange: () => {},
    onRightChange: () => {},
  }

  it('should render without crashing', () => {
    const tree = renderer.create(<RevisionsCompare {...mockPackageHandles} />)
    expect(tree).toBeTruthy()
  })
})
