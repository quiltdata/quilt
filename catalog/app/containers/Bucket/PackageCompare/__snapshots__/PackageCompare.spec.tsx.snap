// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Bucket/PackageCompare/PackageCompare should match snapshot 1`] = `
<div
  className="MuiBox-root MuiBox-root-25"
>
  <div
    className="MuiBox-root MuiBox-root-26"
  >
    <h5
      className="MuiTypography-root MuiTypography-h5"
    >
      <a
        href="/b/test-bucket/packages/test-package"
      >
        test-package
      </a>
       
      revision comparison
    </h5>
  </div>
  <div
    className="MuiBox-root MuiBox-root-27"
  >
    <div
      className="MuiBox-root MuiBox-root-28"
    >
      <div
        className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <table
          className="MuiTable-root"
          role={null}
        >
          <thead
            className="MuiTableHead-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                <h6
                  className="MuiTypography-root MuiTypography-h6"
                >
                  <div
                    className="MuiBox-root MuiBox-root-29"
                  >
                    <div
                      className="MuiBox-root MuiBox-root-30"
                    >
                      hash1
                    </div>
                    <button
                      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={0}
                      type="button"
                    >
                      <span
                        className="MuiIconButton-label"
                      >
                        <span
                          aria-hidden={true}
                          className="material-icons MuiIcon-root MuiIcon-fontSizeSmall"
                        >
                          file_copy
                        </span>
                      </span>
                    </button>
                  </div>
                </h6>
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                <h6
                  className="MuiTypography-root MuiTypography-h6"
                >
                  <div
                    className="MuiBox-root MuiBox-root-31"
                  >
                    <div
                      className="MuiBox-root MuiBox-root-32"
                    >
                      hash2
                    </div>
                    <button
                      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={0}
                      type="button"
                    >
                      <span
                        className="MuiIconButton-label"
                      >
                        <span
                          aria-hidden={true}
                          className="material-icons MuiIcon-root MuiIcon-fontSizeSmall"
                        >
                          file_copy
                        </span>
                      </span>
                    </button>
                  </div>
                </h6>
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root"
              role={null}
            >
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body"
              >
                <a
                  href="/b/test-bucket/packages/test-package/tree/hash1"
                >
                  2023-01-01T00:00:00.000Z
                </a>
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body"
              >
                <a
                  href="/b/test-bucket/packages/test-package/tree/hash2"
                >
                  2023-01-02T00:00:00.000Z
                </a>
              </td>
            </tr>
            <tr
              className="MuiTableRow-root"
              role={null}
            >
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body"
              >
                <p
                  className="MuiTypography-root MuiTypography-body2"
                >
                  Test message 1
                </p>
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body"
              >
                <p
                  className="MuiTypography-root MuiTypography-body2"
                >
                  Test message 2
                </p>
              </td>
            </tr>
            <tr
              className="MuiTableRow-root"
              role={null}
            >
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body"
              >
                <p
                  className="MuiTypography-root MuiTypography-body2"
                >
                  1000 bytes
                </p>
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body"
              >
                <p
                  className="MuiTypography-root MuiTypography-body2"
                >
                  2000 bytes
                </p>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        className="MuiBox-root MuiBox-root-33"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
        >
          Metadata Comparison
        </h6>
        <div
          className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            className="MuiBox-root MuiBox-root-34"
          >
            <div
              data-testid="diff-viewer"
            >
              <div>
                hash1 Metadata
              </div>
              <div>
                hash2 Metadata
              </div>
              <pre>
                {
  "test": "meta1"
}
              </pre>
              <pre>
                {
  "test": "meta2"
}
              </pre>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiBox-root MuiBox-root-35"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
        >
          Entries Comparison
        </h6>
        <div
          className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            className="MuiBox-root MuiBox-root-36"
          >
            <div
              data-testid="diff-viewer"
            >
              <div>
                hash1 Entries
              </div>
              <div>
                hash2 Entries
              </div>
              <pre>
                {
  "file1.txt": {
    "size": 100
  },
  "file2.txt": {
    "size": 200
  }
}
              </pre>
              <pre>
                {
  "file1.txt": {
    "size": 100
  },
  "file2.txt": {
    "size": 200
  }
}
              </pre>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
